File: 1116-maximum-level-sum-of-a-binary-tree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

type Traverse = { node: TreeNode; depth: number };

function maxLevelSum(root: TreeNode | null): number {
  if (!root) return 0;
  const records: number[] = [];
  const traversal: Traverse[] = [{ node: root, depth: 1 }];
  while (traversal[0]) {
    const { node, depth } = traversal.shift() as Traverse;
    records[depth] = (records[depth] ?? 0) + node.val;
    if (node.left) traversal.push({ node: node.left, depth: depth + 1 });
    if (node.right) traversal.push({ node: node.right, depth: depth + 1 });
  }
  return records.reduce(
    (acc, cur, i) => (cur > acc[1] ? [i, cur] : acc),
    [0, -1e9 - 7],
  )[0];
}

File: 1116-maximum-level-sum-of-a-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traversal(TreeNode* node, int lv, vector<int> &sum) {
        if(sum.size() <= lv) sum.push_back(0);
        sum[lv] += node->val;
        if(node->left != NULL) traversal(node->left, lv + 1, sum);
        if(node->right != NULL) traversal(node->right, lv + 1, sum);
    }
        
    int maxLevelSum(TreeNode* root) {
        vector<int> sum;
        traversal(root, 0, sum);
        int ans = 1, maxNow = sum[0];
        for(int i = 2; i <= sum.size(); i++) {
            if(maxNow < sum[i - 1]) {
                maxNow = sum[i - 1];
                ans = i;
            }
        }
        return ans;
    }
};
